docker service inspect --pretty segfaults

**Description**

I created a 2 service via a `docker stack deploy`. Everything worked normally but then a `docker service inspect --pretty SERVICE` returned a SIGSEGV

**Steps to reproduce the issue:**

1. Create an app.yml:
```yaml
version: '3'

services:

  wordpress:
    image: wordpress
    ports:
      - 8080:80
    environment:
      WORDPRESS_DB_PASSWORD: foo!bar

  mysql:
    image: mariadb
    environment:
      MYSQL_ROOT_PASSWORD: foo!bar
```

2. docker stack deploy --compose-file app.yml test
3. docker service inspect --pretty test_mysql

**Describe the results you received:**

```
sam@sam-lab-node-1:~$ docker service inspect --pretty test_mysql
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x7e3f80]

goroutine 1 [running]:
panic(0xab42e0, 0xc4200100b0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
text/template.errRecover(0xc420165080)
	/usr/local/go/src/text/template/exec.go:140 +0x2ad
panic(0xab42e0, 0xc4200100b0)
	/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/docker/docker/cli/command/formatter.(*serviceInspectContext).HasResourceReservations(0xc4203da840, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/command/formatter/service.go:266 +0x10
reflect.Value.call(0xb6b0a0, 0xc4203da840, 0x2a13, 0xb791d9, 0x4, 0x112eab0, 0x0, 0x0, 0x4d9303, 0x1110f60, ...)
	/usr/local/go/src/reflect/value.go:434 +0x5c8
reflect.Value.Call(0xb6b0a0, 0xc4203da840, 0x2a13, 0x112eab0, 0x0, 0x0, 0xc420026988, 0x28, 0x28)
	/usr/local/go/src/reflect/value.go:302 +0xa4
text/template.(*state).evalCall(0xc420165000, 0xb6b0a0, 0xc4203da840, 0x16, 0xb6b0a0, 0xc4203da840, 0x2a13, 0x10f1500, 0xc4203c1860, 0xc42034da68, ...)
	/usr/local/go/src/text/template/exec.go:658 +0x530
text/template.(*state).evalField(0xc420165000, 0xb6b0a0, 0xc4203da840, 0x16, 0xc42034da68, 0x17, 0x10f1500, 0xc4203c1860, 0xc4203995f0, 0x1, ...)
	/usr/local/go/src/text/template/exec.go:549 +0xdb6
text/template.(*state).evalFieldChain(0xc420165000, 0xb6b0a0, 0xc4203da840, 0x16, 0xb6b0a0, 0xc4203da840, 0x16, 0x10f1500, 0xc4203c1860, 0xc4203995e0, ...)
	/usr/local/go/src/text/template/exec.go:520 +0x22f
text/template.(*state).evalFieldNode(0xc420165000, 0xb6b0a0, 0xc4203da840, 0x16, 0xc4203c1860, 0xc4203995f0, 0x1, 0x1, 0x0, 0x0, ...)
	/usr/local/go/src/text/template/exec.go:484 +0x114
text/template.(*state).evalCommand(0xc420165000, 0xb6b0a0, 0xc4203da840, 0x16, 0xc4203c1830, 0x0, 0x0, 0x0, 0x16, 0xb6b0a0, ...)
	/usr/local/go/src/text/template/exec.go:422 +0x866
text/template.(*state).evalPipeline(0xc420165000, 0xb6b0a0, 0xc4203da840, 0x16, 0xc4203c2780, 0x1, 0x0, 0x119)
	/usr/local/go/src/text/template/exec.go:400 +0xee
text/template.(*state).walkIfOrWith(0xc420165000, 0xa, 0xb6b0a0, 0xc4203da840, 0x16, 0xc4203c2780, 0xc4203c1890, 0x0)
	/usr/local/go/src/text/template/exec.go:255 +0xc4
text/template.(*state).walk(0xc420165000, 0xb6b0a0, 0xc4203da840, 0x16, 0x10f15c0, 0xc420325f00)
	/usr/local/go/src/text/template/exec.go:231 +0x2e8
text/template.(*state).walk(0xc420165000, 0xb6b0a0, 0xc4203da840, 0x16, 0x10f1620, 0xc4203c17d0)
	/usr/local/go/src/text/template/exec.go:234 +0x13d
text/template.(*state).walkIfOrWith(0xc420165000, 0xa, 0xb6b0a0, 0xc4203da840, 0x16, 0xc4203c2730, 0xc4203c17d0, 0x0)
	/usr/local/go/src/text/template/exec.go:264 +0x1d0
text/template.(*state).walk(0xc420165000, 0xb6b0a0, 0xc4203da840, 0x16, 0x10f15c0, 0xc4203c60c0)
	/usr/local/go/src/text/template/exec.go:231 +0x2e8
text/template.(*state).walk(0xc420165000, 0xb6b0a0, 0xc4203da840, 0x16, 0x10f1620, 0xc4203a7680)
	/usr/local/go/src/text/template/exec.go:234 +0x13d
text/template.(*Template).execute(0xc420325580, 0x10e6000, 0xc4203b0690, 0xb6b0a0, 0xc4203da840, 0x0, 0x0)
	/usr/local/go/src/text/template/exec.go:189 +0x204
text/template.(*Template).Execute(0xc420325580, 0x10e6000, 0xc4203b0690, 0xb6b0a0, 0xc4203da840, 0x7f3fd2fa2000, 0x0)
	/usr/local/go/src/text/template/exec.go:175 +0x53
github.com/docker/docker/cli/command/formatter.(*Context).contextFormat(0xc4201bd810, 0xc420325580, 0x10ebf80, 0xc4203da840, 0xc420165601, 0xc4203da840)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/command/formatter/formatter.go:91 +0x75
github.com/docker/docker/cli/command/formatter.(*Context).Write.func1(0x10ebf80, 0xc4203da840, 0xc4203da580, 0xc4201656d8)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/command/formatter/formatter.go:115 +0x47
github.com/docker/docker/cli/command/formatter.ServiceInspectWrite.func1(0xc4203a5ee0, 0xc4203a5ee0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/command/formatter/service.go:132 +0x1e8
github.com/docker/docker/cli/command/formatter.(*Context).Write(0xc4201bd810, 0x10ebf80, 0xc420079080, 0xc4203a7590, 0x1, 0xc420165a60)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/command/formatter/formatter.go:117 +0x17d
github.com/docker/docker/cli/command/formatter.ServiceInspectWrite(0x10e6640, 0xc4203033e0, 0xbae942, 0x97d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/command/formatter/service.go:138 +0x209
github.com/docker/docker/cli/command/service.runInspect(0xc4201b94a0, 0xc4203a4a00, 0x1, 0x2, 0xb7b713, 0x6, 0x1, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/command/service/inspect.go:80 +0x2f3
github.com/docker/docker/cli/command/service.newInspectCommand.func1(0xc420357680, 0xc4203a4a00, 0x1, 0x2, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/command/service/inspect.go:35 +0x90
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).execute(0xc420357680, 0xc42000c0d0, 0x2, 0x2, 0xc420357680, 0xc42000c0d0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:646 +0x26d
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc4200738c0, 0xc420073d40, 0xc42032cf50, 0xc420026448)
	/usr/src/docker/.gopath/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:742 +0x377
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4200738c0, 0xc4200738c0, 0x10e7f00)
	/usr/src/docker/.gopath/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:695 +0x2b
main.main()
	/usr/src/docker/.gopath/src/github.com/docker/docker/cmd/docker/docker.go:93 +0xcb
```

**Output of `docker version`:**

```
Client:
 Version:      1.13.0-rc1
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   75fd88b
 Built:        Fri Nov 11 19:33:42 2016
 OS/Arch:      linux/amd64

Server:
 Version:             1.13.0-rc1
 API version:         1.25
 Minimum API version: 1.12
 Go version:          go1.7.3
 Git commit:          75fd88b
 Built:               Fri Nov 11 19:33:42 2016
 OS/Arch:             linux/amd64
 Experimental:        true
```

**Output of `docker info`:**

```
Containers: 2
 Running: 1
 Paused: 0
 Stopped: 1
Images: 16
Server Version: 1.13.0-rc1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 88
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: ew4z6r4bpi3bsdx0v3wmcscrs
 Is Manager: true
 ClusterID: wf62av8eq1uijzzdry2m0qfcg
 Managers: 1
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.31.6.44
 Manager Addresses:
  172.31.6.44:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 8517738ba4b82aff5662c97ca4627e7e4d03b531
runc version: ac031b5bf1cc92239461125f4c1ffb760522bbf2
init version: N/A (expected: v0.13.0)
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.962 GiB
Name: sam-lab-node-1
ID: ZPDA:MPKP:GGN7:UNO6:BFTT:5H7M:O26Z:TPVX:CSIE:TNIF:Z3KF:XTEH
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional notes**

If you omit the `--pretty` option, it works.

**Additional environment details (AWS, VirtualBox, physical, etc.):**

AWS, Debian Jessie

--- a/cli/command/formatter/service.go
+++ b/cli/command/formatter/service.go
@@ -263,6 +263,9 @@ func (ctx *serviceInspectContext) HasResources() bool {
 }
 
 func (ctx *serviceInspectContext) HasResourceReservations() bool {
+	if ctx.Service.Spec.TaskTemplate.Resources == nil || ctx.Service.Spec.TaskTemplate.Resources.Reservations == nil {
+		return false
+	}
 	return ctx.Service.Spec.TaskTemplate.Resources.Reservations.NanoCPUs > 0 || ctx.Service.Spec.TaskTemplate.Resources.Reservations.MemoryBytes > 0
 }
 
@@ -281,6 +284,9 @@ func (ctx *serviceInspectContext) ResourceReservationMemory() string {
 }
 
 func (ctx *serviceInspectContext) HasResourceLimits() bool {
+	if ctx.Service.Spec.TaskTemplate.Resources == nil || ctx.Service.Spec.TaskTemplate.Resources.Limits == nil {
+		return false
+	}
 	return ctx.Service.Spec.TaskTemplate.Resources.Limits.NanoCPUs > 0 || ctx.Service.Spec.TaskTemplate.Resources.Limits.MemoryBytes > 0
 }
 

--- a/integration-cli/docker_cli_swarm_test.go
+++ b/integration-cli/docker_cli_swarm_test.go
@@ -1072,3 +1072,19 @@ func (s *DockerSwarmSuite) TestSwarmManagerAddress(c *check.C) {
 	c.Assert(err, checker.IsNil)
 	c.Assert(out, checker.Contains, expectedOutput)
 }
+
+func (s *DockerSwarmSuite) TestSwarmServiceInspectPretty(c *check.C) {
+	d := s.AddDaemon(c, true, true)
+
+	name := "top"
+	out, err := d.Cmd("service", "create", "--name", name, "--limit-cpu=0.5", "busybox", "top")
+	c.Assert(err, checker.IsNil, check.Commentf(out))
+
+	expectedOutput := `
+Resources:
+ Limits:
+  CPU:		0.5`
+	out, err = d.Cmd("service", "inspect", "--pretty", name)
+	c.Assert(err, checker.IsNil, check.Commentf(out))
+	c.Assert(out, checker.Contains, expectedOutput, check.Commentf(out))
+}

